doc/source/notebooks/tailor/gp_nn.ipynb

```diff
@@ -210,198 +180,132 @@
-    "    \n",
-    "    # Reset inducing (these live in a different space to X, so we need to be careful here)\n",
-    "    ind = np.random.choice(Mnist.X.shape[0], minibatch_size, replace=False)\n",
-    "    \n",
-    "    # Currently a hack is needed due to model initialisation\n",
-    "    feat = KernelSpaceInducingPoints(np.empty((M, fX_dim)))\n",
-    "    #feat = FFeature(Z_0)  # Ideally, we could move the calculation of Z_0\n",
-    "    \n",
-    "    # Build the model \n",
-    "\n",
-    "    lik = gpflow.likelihoods.MultiClass(Mnist.Nclasses)\n",
+    "data_iterator = iter(dataset)\n",
+    "adam_opt = tf.optimizers.Adam(0.001)\n",
     "\n",
-    "    #Z = kmeans2(Mnist.X, M, minit='points')[0]\n",
     "\n",
-    "    model = NN_SVGP(Mnist.X, Mnist.Y, kern, lik, feat=feat, num_latent=Mnist.Nclasses, minibatch_size=minibatch_size)\n",
+    "@tf.function(autograph=False)\n",
+    "def loss_cb(x, y):\n",
+    "    loss_value = -model.elbo(x, y)\n",
+    "    return loss_value\n",
     "\n",
-    "    fZ = model.kern.compute_f(Mnist.X[ind])\n",
-    "    # Z_0 = kmeans2(fZ, M)[0] might fail\n",
-    "    Z_0 = fZ[np.random.choice(len(fZ), M, replace=False)]\n",
-    "    model.feature.Z = Z_0\n",
     "\n",
-    "    # Use GPflow wrappers to train. Note all session handling is done for us\n",
-    "    gpflow.training.AdamOptimizer(0.001).minimize(model, maxiter=ITERATIONS)\n",
+    "@tf.function(autograph=False)\n",
+    "def optimization_step():\n",
+    "    x, y = next(data_iterator)\n",
+    "    func = lambda: loss_cb(x, y)\n",
+    "    adam_opt.minimize(func, var_list=model.trainable_variables)\n",
     "\n",
-    "    # Predictions\n",
-    "    m, v = model.predict_y(Mnist.Xtest)\n",
-    "    preds = np.argmax(m, 1).reshape(Mnist.Ytest.shape)\n",
-    "    correct = preds == Mnist.Ytest.astype(int)\n",
-    "    acc = np.average(correct.astype(float)) * 100.\n",
     "\n",
-    "    print('Accuracy is {:.4f}%'.format(acc))\n",
-    "\n",
-    "gpflow.reset_default_graph_and_session()\n",
-    "ex2b()"
+    "for _ in range(iterations):\n",
+    "    optimization_step()"
    ]
   },
   {
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "## Example 2: A GPflow model on top of a TensorFlow model\n",
-    "Now we'll do things the other way; take a model implemented in TensorFlow and explain how to put a GPflow model on top."
+    "Let's do predictions after training. Don't expect that we will get a good accuracy, as we haven't run training for long enough."
    ]
   },
   {
    "cell_type": "code",
    "execution_count": 7,
-   "metadata": {
-    "ExecuteTime": {
-     "end_time": "2018-06-20T11:40:21.347439Z",
-     "start_time": "2018-06-20T11:25:46.167693Z"
-    }
-   },
+   "metadata": {},
    "outputs": [
     {
      "name": "stdout",
      "output_type": "stream",
      "text": [
-      "acc is 98.5900\n"
+      "Accuracy is 20.6055%\n"
```

Comment:
We probably want to check why performance is so off before we merge this. Was the model trained long enough?

Reply:
@vdutor , run it longer, got 97.2233